File: output/GPT3.5-Turbo/xCodeEval_4000/Java/C/bb78e59b72a929aa7df1b46863a26ebb.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

#define MAX_T 1000000000
#define MODULO 1000000007

typedef struct {
    int type;
    int arg1;
    int arg2;
    int arg3;
} Query;

typedef struct {
    int id;
    int value;
} IdValue;

typedef struct {
    int size;
    IdValue *values;
} IdMap;

typedef struct {
    int size;
    long *sums;
    long *mins;
} SegmentTree;

int readInt() {
    int x;
    scanf("%d", &x);
    return x;
}

void updateIndex(SegmentTree *tree, int index, long value) {
    tree->sums[index] = value;
    tree->mins[index] = fmin(0, value);
}

void updateTreeIndex(SegmentTree *tree, int v, int left, int right, int index, long value) {
    if (left + 1 == right) {
        updateIndex(tree, v, value);
    } else {
        int mid = (left + right) / 2;
        int vLeft = v * 2;
        int vRight = vLeft + 1;

        if (index < mid) {
            updateTreeIndex(tree, vLeft, left, mid, index, value);
        } else {
            updateTreeIndex(tree, vRight, mid, right, index, value);
        }

        tree->sums[v] = tree->sums[vLeft] + tree->sums[vRight];
        tree->mins[v] = fmin(tree->mins[vLeft], tree->sums[vLeft] + tree->mins[vRight]);
    }
}

void updateVertex(SegmentTree *tree, int v, int vLeft, int vRight) {
    tree->sums[v] = tree->sums[vLeft] + tree->sums[vRight];
    tree->mins[v] = fmin(tree->mins[vLeft], tree->sums[vLeft] + tree->mins[vRight]);
}

void updateTree(SegmentTree *tree, Query *queries, int queryCount, IdMap *ids, int *times, int minTime, int maxTime) {
    for (int i = 0; i < queryCount; i++) {
        Query query = queries[i];
        int type = query.type;
        int time = query.arg1;
        int realTime = times[time];
        long speed;

        if (type == 1) {
            speed = query.arg2;

            int timeBefore = ids->values[time - 1].id;
            int realTimeBefore = times[timeBefore];
            long speedBefore = ids->values[time - 1].value;

            updateIndex(tree, timeBefore, speedBefore * (realTime - realTimeBefore));
            updateIndex(tree, time, speed * (times[time + 1] - realTime));
        } else if (type == 2) {
            int timeBefore = ids->values[time - 1].id;
            int realTimeBefore = times[timeBefore];
            long speedBefore = ids->values[time - 1].value;

            updateIndex(tree, timeBefore, speedBefore * (times[time + 1] - realTimeBefore));
            updateIndex(tree, time, 0);
        } else {
            int start = query.arg1;
            int end = query.arg2;
            long value = query.arg3;

            int foundIndex = -1;
            for (int j = start; j <= end; j++) {
                if (tree->sums[j] + tree->mins[j] <= value) {
                    foundIndex = j;
                    break;
                }
            }

            if (foundIndex == -1) {
                printf("-1\n");
            } else {
                value = tree->sums[foundIndex];

                int time = foundIndex;
                int realTime = times[time];
                long speed = ids->values[time].value;

                int timeAfter = fmin(end, ids->values[time + 1].id);
                int realTimeAfter = times[timeAfter];

                long delta = realTimeAfter - realTime;
                long total = delta * speed;

                if (value + total > 0) {
                    printf("-1\n");
                } else {
                    double badTime = realTime + (double)value / fabs(speed);
                    printf("%f\n", badTime);
                }
            }
        }
    }
}

void solve() {
    int q = readInt();
    Query *queries = (Query *)malloc(q * sizeof(Query));

    for (int i = 0; i < q; i++) {
        int type = readInt();
        if (type == 1) {
            queries[i].type = 1;
            queries[i].arg1 = readInt();
            queries[i].arg2 = readInt();
        } else if (type == 2) {
            queries[i].type = 2;
            queries[i].arg1 = readInt();
        } else {
            queries[i].type = 3;
            queries[i].arg1 = readInt();
            queries[i].arg2 = readInt();
            queries[i].arg3 = readInt();
        }
    }

    IdMap *mapper = (IdMap *)malloc(sizeof(IdMap));
    mapper->size = 0;
    mapper->values = (IdValue *)malloc((q + 2) * sizeof(IdValue));
    mapper->values[mapper->size].id = 0;
    mapper->values[mapper->size].value = 0;
    mapper->size++;

    for (int i = 0; i < q; i++) {
        Query query = queries[i];
        if (query.type == 1) {
            mapper->values[mapper->size].id = query.arg1;
            mapper->values[mapper->size].value = 0;
            mapper->size++;
        } else if (query.type == 3) {
            mapper->values[mapper->size].id = query.arg1;
            mapper->values[mapper->size].value = 0;
            mapper->size++;
            mapper->values[mapper->size].id = query.arg2;
            mapper->values[mapper->size].value = 0;
            mapper->size++;
        }
    }

    IdMap *ids = (IdMap *)malloc(sizeof(IdMap));
    ids->size = mapper->size;
    ids->values = (IdValue *)malloc(ids->size * sizeof(IdValue));

    for (int i = 0; i < ids->size; i++) {
        ids->values[i].id = mapper->values[i].id;
        ids->values[i].value = i;
    }

    int *times = (int *)malloc(ids->size * sizeof(int));
    for (int i = 0; i < ids->size; i++) {
        times[i] = ids->values[i].id;
    }

    int minTime = ids->values[0].id;
    int maxTime = ids->values[ids->size - 1].id;

    SegmentTree *tree = (SegmentTree *)malloc(sizeof(SegmentTree));
    tree->size = ids->size;
    tree->sums = (long *)malloc((4 * tree->size) * sizeof(long));
    tree->mins = (long *)malloc((4 * tree->size) * sizeof(long));

    for (int i = 0; i < tree->size; i++) {
        tree->sums[i] = 0;
        tree->mins[i] = 0;
    }

    updateIndex(tree, 0, 0);
    updateIndex(tree, tree->size - 1, 0);

    for (int i = 0; i < q; i++) {
        Query query = queries[i];
        int type = query.type;
        if (type == 1) {
            int time = query.arg1;
            int realTime = times[time];
            long speed = query.arg2;

            int timeBefore = ids->values[time - 1].id;
            int realTimeBefore = times[timeBefore];
            long speedBefore = ids->values[time - 1].value;

            updateIndex(tree, timeBefore, speedBefore * (realTime - realTimeBefore));
            updateIndex(tree, time, speed * (times[time + 1] - realTime));
        } else if (type == 2) {
            int time = query.arg1;
            int realTime = times[time];

            int timeBefore = ids->values[time - 1].id;
            int realTimeBefore = times[timeBefore];
            long speedBefore = ids->values[time - 1].value;

            updateIndex(tree, timeBefore, speedBefore * (times[time + 1] - realTimeBefore));
            updateIndex(tree, time, 0);
        } else {
            int start = query.arg1;
            int end = query.arg2;
            long value = query.arg3;

            int foundIndex = -1;
            for (int j = start; j <= end; j++) {
                if (tree->sums[j] + tree->mins[j] <= value) {
                    foundIndex = j;
                    break;
                }
            }

            if (foundIndex == -1) {
                printf("-1\n");
            } else {
                value = tree->sums[foundIndex];

                int time = foundIndex;
                int realTime = times[time];
                long speed = ids->values[time].value;

                int timeAfter = fmin(end, ids->values[time + 1].id);
                int realTimeAfter = times[timeAfter];

                long delta = realTimeAfter - realTime;
                long total = delta * speed;

                if (value + total > 0) {
                    printf("-1\n");
                } else {
                    double badTime = realTime + (double)value / fabs(speed);
                    printf("%f\n", badTime);
                }
            }
        }
    }

    free(queries);
    free(mapper->values);
    free(mapper);
    free(ids->values);
    free(ids);
    free(times);
    free(tree->sums);
    free(tree->mins);
    free(tree);
}

int main() {
    int q = readInt();
    for (int i = 0; i < q; i++) {
        solve();
    }
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_4000/Java/Go/bb78e59b72a929aa7df1b46863a26ebb.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
)

const MODULO = 1000000007
const MAX_T = 1000000000

type Query struct {
	Type int
	Args []int
}

type SegmentTree struct {
	size  int
	sums  []int64
	mins  []int64
	value int64
	start int
	end   int
}

func NewSegmentTree(n int) *SegmentTree {
	treeSize := n << 2
	return &SegmentTree{
		size: n,
		sums: make([]int64, treeSize),
		mins: make([]int64, treeSize),
	}
}

func (tree *SegmentTree) UpdateIndex(index int, value int64) {
	tree.value = value
	tree.start = index
	tree.updateTreeIndex(1, 0, tree.size)
}

func (tree *SegmentTree) updateTreeIndex(v, left, right int) {
	if left+1 == right {
		tree.sums[v] = tree.value
		tree.mins[v] = min(0, tree.value)
	} else {
		mid := (left + right) >> 1
		vLeft := v << 1
		vRight := vLeft + 1

		if tree.start < mid {
			tree.updateTreeIndex(vLeft, left, mid)
		} else {
			tree.updateTreeIndex(vRight, mid, right)
		}

		tree.updateVertex(v, vLeft, vRight)
	}
}

func (tree *SegmentTree) updateVertex(v, vLeft, vRight int) {
	tree.sums[v] = tree.sums[vLeft] + tree.sums[vRight]
	tree.mins[v] = min(tree.mins[vLeft], tree.sums[vLeft]+tree.mins[vRight])
}

func (tree *SegmentTree) GetInterval(start, end int, value int64) int {
	tree.start = start
	tree.end = end
	return tree.getTreeInterval(1, 0, tree.size)
}

func (tree *SegmentTree) getTreeInterval(v, left, right int) int {
	if tree.start <= left && right <= tree.end {
		if value+tree.mins[v] <= 0 {
			return tree.findTreeIndex(v, left, right)
		} else {
			value += tree.sums[v]
		}
	} else {
		mid := (left + right) >> 1
		vLeft := v << 1
		vRight := vLeft + 1

		if tree.start < mid {
			tree.getTreeInterval(vLeft, left, mid)
		}
		if mid < tree.end {
			tree.getTreeInterval(vRight, mid, right)
		}
	}
	return -1
}

func (tree *SegmentTree) findTreeIndex(v, left, right int) int {
	if left+1 == right {
		return left
	} else {
		mid := (left + right) >> 1
		vLeft := v << 1
		vRight := vLeft + 1

		if value+tree.mins[vLeft] <= 0 {
			return tree.findTreeIndex(vLeft, left, mid)
		} else {
			value += tree.sums[vLeft]
			return tree.findTreeIndex(vRight, mid, right)
		}
	}
}

type IdMap struct {
	mapping map[int]int
}

func NewIdMap() *IdMap {
	return &IdMap{mapping: make(map[int]int)}
}

func (idMap *IdMap) Register(key int) int {
	if id, exists := idMap.mapping[key]; exists {
		return id
	}
	id := len(idMap.mapping)
	idMap.mapping[key] = id
	return id
}

type SortedIdMapper struct {
	values []int
}

func NewSortedIdMapper() *SortedIdMapper {
	return &SortedIdMapper{values: []int{}}
}

func (mapper *SortedIdMapper) AddValue(value int) {
	mapper.values = append(mapper.values, value)
}

func (mapper *SortedIdMapper) Build() *IdMap {
	sort.Ints(mapper.values)
	idMap := NewIdMap()
	uniqueValues := []int{}

	for i, value := range mapper.values {
		if i == 0 || mapper.values[i-1] != value {
			idMap.Register(value)
			uniqueValues = append(uniqueValues, value)
		}
	}

	mapper.values = uniqueValues
	return idMap
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var q int
	fmt.Fscan(reader, &q)
	queries := make([]Query, q)

	for i := 0; i < q; i++ {
		var typ int
		fmt.Fscan(reader, &typ)
		if typ == 1 {
			var a, b int
			fmt.Fscan(reader, &a, &b)
			queries[i] = Query{Type: typ, Args: []int{a, b}}
		} else if typ == 2 {
			var a int
			fmt.Fscan(reader, &a)
			queries[i] = Query{Type: typ, Args: []int{a}}
		} else {
			var a, b, c int
			fmt.Fscan(reader, &a, &b, &c)
			queries[i] = Query{Type: typ, Args: []int{a, b, c}}
		}
	}

	mapper := NewSortedIdMapper()
	mapper.AddValue(0)
	mapper.AddValue(MAX_T + 1)

	for _, query := range queries {
		mapper.AddValue(query.Args[0])
		if query.Type == 3 {
			mapper.AddValue(query.Args[1])
		}
	}

	ids := mapper.Build()
	times := make([]int, len(mapper.values))
	for i, v := range mapper.values {
		times[i] = v
	}

	minTime := ids.mapping[0]
	maxTime := ids.mapping[MAX_T+1]
	timeSpeeds := make(map[int]int)

	timeSpeeds[minTime] = 0
	timeSpeeds[maxTime] = 0

	for i := range queries {
		queries[i].Args[0] = ids.mapping[queries[i].Args[0]]
		if queries[i].Type == 3 {
			queries[i].Args[1] = ids.mapping[queries[i].Args[1]]
		}
	}

	tree := NewSegmentTree(len(ids.mapping))

	for _, query := range queries {
		switch query.Type {
		case 1:
			time := query.Args[0]
			realTime := times[time]
			speed := int64(query.Args[1])

			entryBefore := timeSpeeds[time-1]
			timeAfter := timeSpeeds[time+1]
			realTimeAfter := times[timeAfter]

			timeBefore := entryBefore
			realTimeBefore := times[timeBefore]
			speedBefore := int64(timeSpeeds[timeBefore])

			tree.UpdateIndex(timeBefore, speedBefore*(realTime-realTimeBefore))
			tree.UpdateIndex(time, speed*(realTimeAfter-realTime))

			timeSpeeds[time] = int(speed)

		case 2:
			time := query.Args[0]
			realTime := times[time]

			entryBefore := timeSpeeds[time-1]
			timeAfter := timeSpeeds[time+1]
			realTimeAfter := times[timeAfter]

			timeBefore := entryBefore
			realTimeBefore := times[timeBefore]
			speedBefore := int64(timeSpeeds[timeBefore])

			tree.UpdateIndex(timeBefore, speedBefore*(realTimeAfter-realTimeBefore))
			tree.UpdateIndex(time, 0)

			delete(timeSpeeds, time)

		case 3:
			start := query.Args[0]
			end := query.Args[1]
			value := int64(query.Args[2])

			foundIndex := tree.GetInterval(start, end, value)
			if foundIndex == -1 {
				fmt.Fprintln(writer, -1)
			} else {
				value = tree.value

				time := foundIndex
				realTime := times[time]

				if value == 0 {
					fmt.Fprintln(writer, realTime)
				} else {
					speed := int64(timeSpeeds[time])
					timeAfter := timeSpeeds[time+1]
					realTimeAfter := times[timeAfter]

					delta := realTimeAfter - realTime
					total := delta * speed

					if value+total > 0 {
						fmt.Fprintln(writer, -1)
					} else {
						badTime := float64(realTime) + float64(value)/math.Abs(float64(speed))
						fmt.Fprintln(writer, badTime)
					}
				}
			}
		}
	}
}
